documentation/Intrinsics.md

```diff
@@ -683,3 +685,103 @@ CACHESIZE, EOF, FP_CLASS, INT_PTR_KIND, ISNAN, LOC
+F18 automatically detects if these types match an f18 scalar type. If so,
+folding of the intrinsic functions will be possible for the related f18 scalar type,
+otherwise an error message will be produced by f18 when attempting to fold related intrinsic functions.
+
+| C/C++ Host Type | Intrinsic Functions with Host Standard C++ Library Based Folding Support |
+| --- | --- |
+| float, double and long double | ACOS, ACOSH, ASINH, ATAN, ATAN2, ATANH, COS, COSH, ERF, ERFC, EXP, GAMMA, HYPOT, LOG, LOG10, LOG_GAMMA, MOD, SIN, SQRT, SINH, SQRT, TAN, TANH |
+| std::complex for float, double and long double| ACOS, ACOSH, ASIN, ASINH, ATAN, ATANH, COS, COSH, EXP, LOG, SIN, SINH, SQRT, TAN, TANH |
+
+On top of the default usage of C++ standard library functions for folding described
+in the table above, it is possible to compile f18 evaluate library with
+[libpgmath](https://github.com/flang-compiler/flang/tree/master/runtime/libpgmath)
+so that it can be used for folding. To do so, one must have a compiled version
+of the libpgmath library available on the host and add
+`-DLIBPGMATH_DIR=<path to the compiled shared libpgmath library>` to the f18 cmake command.
+
+Libpgmath comes with real and complex functions that replace C++ standard library
+float and double functions to fold all the intrinsic functions listed in the table above.
+It has no long double versions. If the host long double matches an f18 scalar type,
+C++ standard library functions will still be used for folding expressions with this scalar type.
+Libpgmath adds the possibility to fold the following functions for f18 real scalar
+types related to host float and double types.
+
+| C/C++ Host Type | Additional Intrinsic Function Folding Support with Libpgmath (Optional) |
+| --- | --- |
+|float and double| BESSEL_J0, BESSEL_J1, BESSEL_JN (elemental only), BESSEL_Y0, BESSEL_Y1, BESSEL_Yn (elemental only), ERFC_SCALED |
+
+Lipgmath comes in three variants (precise, relaxed and fast). So far, only the
+precise version is used for intrinsic function folding in f18. It guarantees the greatest numerical precision.
+
+### Intrinsic Functions with Missing Folding Support
+The following intrinsic functions are allowed in constant expressions but f18
+is not yet able to fold them. Note that there might be constraints on the arguments
+so that these intrinsics can be used in constant expressions (see section 10.1.12 of Fortran 2018 standard).
+
+ALL, ACHAR, ADJUSTL, ADJUSTR, ANINT, ANY, BESSEL_JN (transformational only),
+BESSEL_YN (transformational only), BTEST, CEILING, CHAR, COUNT, CSHIFT, DOT_PRODUCT,
+DIM (REAL only), DOT_PRODUCT, EOSHIFT, FINDLOC, FLOOR, FRACTION, HUGE, IACHAR, IALL,
+IANY, IPARITY, IBITS, ICHAR, IMAGE_STATUS, INDEX, ISHFTC, IS_IOSTAT_END,
+IS_IOSTAT_EOR, LBOUND, LEN_TRIM, LGE, LGT, LLE, LLT, LOGICAL, MATMUL, MAX, MAXLOC,
+MAXVAL, MERGE, MIN, MINLOC, MINVAL, MOD (INTEGER only), MODULO, NEAREST, NINT,
```

Comment:
What is the plan for MOD (Integer) ?

Reply:
Actually, It is already implemented on F18 scalar integers as `DivideSigned` in [lib/evaluate/integer.h](https://github.com/flang-compiler/f18/blob/master/lib/evaluate/integer.h). I just did not plug it yet in the folding framework because the main purpose of this effort was to design the framework for supporting intrinsic folding and then implement enough of the framework to evaluate the design.
The remainder of the work will be done later when we are closer to executing code.